File: stepik-union/src/main/java/org/stepik/plugin/projectView/NavBarModelExtensionUtils.java
package org.stepik.plugin.projectView;

import com.intellij.ide.projectView.PresentationData;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiDirectory;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.jetbrains.tmp.learning.StepikProjectManager;
import com.jetbrains.tmp.learning.courseFormat.Course;
import org.jetbrains.annotations.Nullable;

import static org.stepik.plugin.utils.PresentationDataUtils.isVisibleDirectory;
import static org.stepik.plugin.utils.PresentationDataUtils.isVisibleFile;
import static org.stepik.plugin.utils.PresentationDataUtils.updatePresentationData;

/**
 * @author meanmail
 */
class NavBarModelExtensionUtils {
    @Nullable
    static String getPresentableText(@Nullable final Object object) {
        if (object instanceof Project) {
            Project project = (Project) object;
            StepikProjectManager stepikProjectManager = StepikProjectManager.getInstance(project);
            Course course = stepikProjectManager.getCourse();
            if (course == null)
                return null;
            return course.getName();
        }

        if (object instanceof PsiDirectory) {
            PsiDirectory psiDirectory = (PsiDirectory) object;
            PresentationData data = new PresentationData();
            updatePresentationData(data, psiDirectory);
            String text = data.getPresentableText();
            if (text != null)
                return text;
        }

        return null;
    }

    @Nullable
    static PsiElement adjustElement(final PsiElement psiElement) {
        Project project = psiElement.getProject();
        StepikProjectManager stepikProjectManager = StepikProjectManager.getInstance(project);
        Course course = stepikProjectManager.getCourse();
        if (course == null)
            return psiElement;

        if (psiElement instanceof PsiDirectory) {
            if (!isVisibleDirectory((PsiDirectory) psiElement))
                return null;
        } else if (psiElement instanceof PsiFile) {
            if (!isVisibleFile((PsiFile) psiElement))
                return null;
        }

        return psiElement;
    }
}

File: stepik-core/src/main/kotlin/org/stepik/core/projectView/NavBarModelExtensionUtils.kt
package org.stepik.core.projectView

import com.intellij.ide.projectView.PresentationData
import com.intellij.openapi.components.ServiceManager.getService
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiDirectory
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import org.stepik.core.ProjectManager
import org.stepik.core.utils.isVisibleDirectory
import org.stepik.core.utils.isVisibleFile
import org.stepik.core.utils.updatePresentationData


object NavBarModelExtensionUtils {
    fun getPresentableText(any: Any?): String? {
        if (any is Project) {
            val projectManager = getService(any, ProjectManager::class.java)
            return projectManager.projectRoot?.name
        }

        if (any is PsiDirectory) {
            val data = PresentationData()
            updatePresentationData(data, any)
            return data.presentableText
        }

        return null
    }

    fun adjustElement(psiElement: PsiElement): PsiElement? {
        val projectManager = getService(psiElement.project, ProjectManager::class.java)
        projectManager?.projectRoot ?: return psiElement

        when (psiElement) {
            is PsiDirectory -> if (!psiElement.isVisibleDirectory()) return null
            is PsiFile -> if (!psiElement.isVisibleFile()) return null
        }

        return psiElement
    }
}